src/lib/updating-element.ts

```diff
@@ -0,0 +1,501 @@
+   */
+  protected getProperty(name: string) {
+    return this._props[name];
+  }
+
+  /**
+   * Sets the property `name` to the given `value`. This method is used
+   * to set property values of registered properties.
+   * Setting a property value calls `invalidate` which asynchronously updates
+   * the element. If a property value is set inside the `update` method,
+   * it does not trigger `invalidate`.
+   */
+  // tslint:disable-next-line no-any
+  protected setProperty(name: string, value: any) {
+    const old = this._props[name];
+    const ctor = (this.constructor as typeof UpdatingElement);
+    if (ctor.propertyShouldInvalidate(name, value, old)) {
+      // track old value when changing.
+      if (!this._changedProps) {
+        this._changedProps = {};
+      }
+      if (!(name in this._changedProps)) {
+        this._changedProps[name] = this._props[name];
+      }
+      this._props[name] = value;
+      this._propertyToAttribute(name, value);
+      this.invalidate();
+    }
+  }
+
+  private _propertyToAttribute(name: string, value: any) {
+    const ctor = (this.constructor as typeof UpdatingElement);
+    const attrValue = ctor.propertyValueToAttribute(name, value);
+    if (attrValue !== undefined) {
+      const attr = ctor.attributeNameForProperty(name);
+      if (attr !== undefined) {
+        // track the attr name/value being set to be able to avoid
+        // reflecting back to the property setter via attributeChangedCallback.
+        this._serializingInfo = [attr, attrValue];
+        if (attrValue === null) {
+          this.removeAttribute(attr);
+        } else {
+          this.setAttribute(attr, attrValue);
+        }
+        this._serializingInfo = null;
+      }
+    }
+  }
+
+  private _attributeToProperty(name: string, value: string) {
+    // Use tracking info to avoid deserializing attribute value if it was
+    // just set from a property setter.
+    if (!this._serializingInfo ||
+        (this._serializingInfo[0] !== name && this._serializingInfo[1] !== value)) {
+      const ctor = (this.constructor as typeof UpdatingElement);
+      const propName = ctor._attributeToPropertyMap[name];
+      this[propName as keyof this] = ctor.propertyValueFromAttribute(propName, value);
+    }
+  }
+
+  /**
+   * Invalidates the element causing it to asynchronously update regardless
+   * of whether or not any property changes are pending. This method is
+   * automatically called when any registered property changes. Returns a Promise
+   * that resolves when the element has finished updating.
+   */
+  async invalidate() {
+    // Do not re-queue validation if already invalid (pending) or currently updating.
+    if (this._isPendingUpdate()) {
+      return this._validatePromise;
+    }
+    this._validationState = ValidationState.Invalid;
+    // Make a new validate promise and put the resolver on the stack.
+    this._validatePromise = new Promise((resolve) => this._validateResolvers.push(resolve));
```

Comment:
Why is there a stack of resolvers? Could use some docs around it's relation to reentrant `invalidate` calls.

Reply:
Gone.